### DISCUSSION TITLE
Run hurl in browser/wasm for interactive testing

### DISCUSSION THREAD

**Discussion by infogulch** [Orange-OpenSource/hurl#2613] [Ideas]
hurl has been great for running automated tests against my server. I think it would be neat if hurl could be compiled to wasm and run in the browser so developers can experiment with and author hurl requests interactively.

I might expect a wasm module interface that references an external function that provides access to the fetch api and exposes a function that can execute a hurl request string. A usable interactive interface could be built up around such a primitive.

Aside from features that would have to be disabled because wasm doesn't have access, like reading from local files, it seems like this could be possible in theory.

Thoughts?

**Comment by jcamiel** [comment]
Hi @infogulch,

It's really interesting, but I think it'll be hard to implement. Right now, we've two major C dependencies that will be difficult to port on wasm: libcurl and libxml. As you're certainly aware, Hurl is heavily relying on libcurl to do HTTP exchanges and even is exposing many of its APIs. It will be difficult to have a wasm backend that replicate it (HTTP/3 for instance).

That said, nothing prevents you to try to compile Hurl on wasm, if we can provide some help, don't hesitate to ask us!

**Reply by infogulch** [reply]
Alas the Chrome TcpSocket api from 2012 never got traction 😆.

Yes I imagine even it did compile its functionality would be severely limited: requests would have to filter through the fetch api so no direct control over connections (it's whatever the browser decides), no filesystem so file uploads wouldn't work, etc etc. Even with these limitations it could still be useful for some narrow "typical" case, leaving edge cases to the normal hurl cli in CI or something.

I might try to compile at some point. Leaving some breadcrumbs for myself:

* Someone tried to build libxml2 to wasm 5 years ago: https://github.com/matiasinsaurralde/wasm-libxml2
* Someone tried to build curl to wasm this year but ran into some issues: https://github.com/curl/curl/issues/13018

**Reply by SilenLoc** [reply]
Hi @infogulch,

I had the same idea ^^

 I tried a bit around, hurl core was actually once wasm compatible. I built a ui with egui for it. Since [libxl2 got into core ](https://github.com/Orange-OpenSource/hurl/discussions/2025) it fell apart.

If someone manages to implement libxml2 in pure rust, that would be make it possible to have at least the parser available in web.

Now I rely on the capabilities delivered by a [server](https://lunchtimecode.github.io/hurlalot/) (only parsing the file for now) on shuttle.rs which has terrible uptime for my plan.

## However

A cool way to build a portable ui is "Tauri", the "backend" runs native so you can incorperate Hurl completly while the frontend does not need any dependency on Hurl. 

Of course this does not run in the web but you can write the frontend in any web compatible language. 

This is my goal right now, building a UI on a Hurl Backend with Tauri.

**Comment by lambrospetrou** [comment]
It might be considered an advertisement, but it's free and can be used without signup so suggesting it here :) If the maintainers feel it's not suitable, please delete my comment or let me know to do it myself.

You can use the Open Editor in https://www.skybear.net/scripts/open-editor/ to interactively test your Hurl scripts using an online UI and an online execution platform.

You can pretty much use most of the Hurl options as well!

The free open editor does not allow you to persist response bodies long-term, and is more rate limited than the actual signed in accounts, but it does the job for interactive script authoring. With signed in accounts, you get historical runs listing, and also response bodies persistence for later debugging.

Try it out and let me know what you think. 👌
